src/pymor/reductors/loewner.py

```diff
@@ -0,0 +1,354 @@
+# Copyright pyMOR developers and contributors. All rights reserved.
+# License: BSD 2-Clause License (https://opensource.org/licenses/BSD-2-Clause)
+
+import numpy as np
+import scipy.linalg as spla
+
+from pymor.core.base import BasicObject
+from pymor.models.iosys import LTIModel
+from pymor.models.transfer_function import TransferFunction
+from pymor.tools.random import new_rng
+
+
+class LoewnerReductor(BasicObject):
+    """Reductor based on Loewner interpolation framework.
+
+    The reductor implements interpolation based on the Loewner framework as in :cite:`ALI17`.
+
+    Parameters
+    ----------
+    s
+        |Numpy Array| of shape (n,) containing the frequencies.
+    Hs
+        |Numpy Array| of shape (n, p, m) for MIMO systems with p outputs and m inputs or
+        |Numpy Array| of shape (n,) for SISO systems where the |Numpy Arrays| resemble the transfer
+        function samples. Alternatively, |TransferFunction| or `Model` with `transfer_function`
+        attribute.
+    partitioning
+        `str` or `tuple` of length 2. Strings can either be 'even-odd' or 'half-half' defining
+        the splitting rule. A user-defined partitioning can be defined by passing a tuple of the
+        left and right indices. Defaults to `even-odd`.
+    ordering
+        The ordering with respect to which the splitting rule is executed. Can be either
+        'magnitude', 'random' or 'regular'. Defaults to 'regular'.
+    conjugate
+        Whether to guarantee realness of reduced |LTIModel| by keeping complex conjugates in the
+        same partitioning or not. If `True` will automatically generate conjugate data if necessary.
+    mimo_handling
+        Option indicating how to treat MIMO systems. Can be:
+
+        - `'random'` for using random tangential directions.
+        - `'full'` for fully interpolating all input-output pairs.
+        - Tuple `(ltd, rtd)` where `ltd` corresponds to left and `rtd` to right tangential
+          directions.
+    """
+
+    def __init__(self, s, Hs, partitioning='even-odd', ordering='regular', conjugate=True, mimo_handling='random'):
+        self.__auto_init(locals())
+        self.loewner_svds = None
+
+    def reduce(self, r=None, tol=1e-7):
+        """Reduce using Loewner framework.
+
+        Parameters
+        ----------
+        r
+            Integer for target order of reduced model. If an interpolant with order less than r
+            exists then the output will have the minimal order of an interpolant. Otherwise, the
+            output will be an |LTIModel| with order r. If `None` the order of the reduced model will
+            be the minimal order of an interpolant.
+        tol
+            Truncation tolerance for rank of Loewner matrices.
+
+        Returns
+        -------
+        rom
+            Reduced |LTIModel|.
+        """
+        L, Ls, V, W = loewner_quadruple(self.s, self.Hs, partitioning=self.partitioning, ordering=self.ordering,
```

Comment:
Wouldn't it be better to also cache the Loewner quadruple since it can take some time to construct?